File: crypto/argon2_conformance_test.go
package crypto

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"
)

type ConformanceVector struct {
	Password    string `json:"password"`
	SaltHex     string `json:"salt_hex"`
	MemoryKiB   uint32 `json:"m_kib"`
	Time        uint32 `json:"t"`
	Parallel    uint8  `json:"p"`
	KeyLen      uint32 `json:"dk"`
	ExpectedHex string `json:"expected_hex"`
}

func TestArgon2ConformanceFixture(t *testing.T) {
	raw, err := os.ReadFile("testdata/argon2-conformance-vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	var vector ConformanceVector
	if err := json.Unmarshal(raw, &vector); err != nil {
		t.Fatal(err)
	}
	salt, err := hex.DecodeString(vector.SaltHex)
	if err != nil {
		t.Fatal(err)
	}
	derived, err := DeriveArgon2IDKey(
		[]byte(vector.Password),
		salt,
		vector.KeyLen,
		vector.MemoryKiB,
		vector.Time,
		vector.Parallel,
	)
	if err != nil {
		t.Fatal(err)
	}
	if hex.EncodeToString(derived) != vector.ExpectedHex {
		t.Fatal("Argon2id result does not match pinned fixture")
	}
}

File: client/static/js/src/__tests__/argon2-conformance.test.ts
/**
 * Cross-Language Argon2id Conformance Test (finding E1b)
 *
 * Consumes a pinned fixture shared with Go.
 */

import { describe, test, expect } from 'bun:test';
import { deriveKeyArgon2id, toHex } from '../crypto/primitives';
import conformanceFixture from '../../../../../crypto/testdata/argon2-conformance-vectors.json';

describe('cross-language Argon2id conformance', () => {
  test('derives identical key as Go implementation', async () => {
    // Decode salt from hex
    const saltBytes = new Uint8Array(
      conformanceFixture.salt_hex.match(/.{1,2}/g)!.map((byte) => parseInt(byte, 16))
    );

    const result = await deriveKeyArgon2id({
      password: conformanceFixture.password,
      salt: saltBytes,
      params: {
        memoryCost: conformanceFixture.m_kib,
        timeCost: conformanceFixture.t,
        parallelism: conformanceFixture.p,
        keyLength: conformanceFixture.dk,
        variant: 2,
      },
    });

    const derivedHex = toHex(result.key);

    expect(derivedHex).toBe(conformanceFixture.expected_hex);
  }, 30000); // Allow up to 30s for memory-hard Argon2id derivation
});
